packages/react-devtools-shared/src/backend/renderer.js

```diff
@@ -1503,6 +1679,117 @@ export function attach(
     return changedKeys;
   }
 
+  function canObjectBeMemoized(prev: any, next: any): boolean {
+    if (prev == null || next == null) {
+      return false;
+    }
+
+    // We can't report anything meaningful for hooks changes.
+    if (isHooksChange(next)) {
+      return false;
+    }
+
+    const keys = new Set([...Object.keys(prev), ...Object.keys(next)]);
+    return Array.from(keys).every(key => isEqual(prev[key], next[key]));
+  }
+
+  function isFunctionOrComponent(value: any) {
+    return typeof value === 'function' || isValidElement(value);
+  }
+
+  /**
+   * Scan through a fiber's parent to find function & component props that
+   * weren't passed through props or created from a callback/memo.
+   */
+  function getNonMemoizedProps(
+    fiber: Fiber,
+    changedPropKeys: Array<string> | null,
+  ): Array<string> | null {
+    const debugOwner = fiber._debugOwner;
+    if (
+      changedPropKeys == null ||
+      debugOwner == null ||
+      // Ignore memo parents since they just pass the props through.
+      getTypeSymbol(debugOwner.type) === MEMO_SYMBOL_STRING
+    ) {
+      return null;
+    }
+
+    // Filter props that are passed through from the parent
+    const createdProps = changedPropKeys
+      // Ignore the children prop.
+      .filter(prop => prop !== 'children')
+      .filter(
+        prop => debugOwner.memoizedProps[prop] !== fiber.memoizedProps[prop],
+      )
+      // For now, only filter functions and elements.
+      .filter(prop => isFunctionOrComponent(fiber.memoizedProps[prop]));
+
+    if (createdProps.length === 0) return null;
+
+    const elementType = getElementTypeForFiber(debugOwner);
+
+    let memoValues = [];
+    if (elementType === FunctionComponent) {
+      memoValues = getMemoValues(debugOwner);
+    } else if (elementType === ClassComponent) {
+      memoValues = getStaticClassValues(debugOwner);
```

Comment:
I don't think I understand the purpose of this or how it's used. Can you explain why static class methods are relevant to re-rendering?

Static methods don't cause a component to re-render, only props/state.

Reply:
I am still quite confused.

On the one hand, we're talking about `props` which get passed to a React component and cause it to render (or not). This code is looking at properties on the `prototype` which aren't relevant to re-rendering at all.

Can you please clarify the intent of this code? An example might be helpful.

Reply:
Here is an example sandbox: https://codesandbox.io/s/react-playground-forked-707py

The goal is to find props that aren't being memoized from a class component parent. In this case, we want to detect callback4 as being non memoized:

```
      <Child
        callback1={this._callbackVar}
        callback2={this.state.callback}
        callback3={this._callbackFn}
        callback4={() => {}}
      />
```

Callback1 is considered static/memoized because its a class variable and won't change
Callback2 is in the state object and should preserve reference
Callback3 is a class function, so its static as well

![image](https://user-images.githubusercontent.com/2875524/143069005-a793b5aa-4bb1-4619-836e-3fbece061b63.png)

Reply:
Right, but the way we should be checking for function equality is by comparing pending props to memoized props. We shouldn't need to know or care about the fact that these props are static b'c they're on a prototype. (They could be static b'c they're module scope for instance. Or even b'c they're passed down from a component that didn't re-render.)

Feels wrong to me that class is being considered here at all.

Reply:
> Right, but the way we should be checking for function equality is by comparing pending props to memoized props. 

This only tells us that we had a new function reference; not if it was memoized or not. We would never be able to catch a component that is being forced to re-render because a function is declared inline.

If a function reference didn't change, this code path isn't executed. Otherwise, we need to figure out if this new function reference was memoized or not. For functional components its easy enough to figure out. For class components, I don't see how else to do this.

Reply:
> This only tells us that we had a new function reference; not if it was memoized or not.

I don't see why a function being memoized _as a class property_ is any different than one being memoized by a library like `memoize-one` or `lodash/memoized`. The relevant bit of information is whether or not the function changes between renders. Up to you (as the programmer) to take that information and say "ah, this changed unnecessarily" or "I could memoize this!"

What am I missing

Reply:
I think we are missing a couple things. 

> Up to you (as the programmer) to take that information and say "ah, this changed unnecessarily" or "I could memoize this!"

This is the problem I'm trying to solve. Currently to process this information we need to dig into the source code and see if the property was memoized (e.g. `useCallback`, class property, class protoype, etc). This is inefficient and slow.

> I don't see why a function being memoized as a class property is any different than one being memoized by a library like memoize-one or lodash/memoized.

It's not. But class prototype and class properties are a good start for finding these memoized values that we can ignore when doing the performance investigation.